### ISSUE TITLE
Batched updates break interaction tracing for mounts

### ISSUE THREAD

**Issue by bvaughn** [facebook/react#15566]
This test works:
```js
let interaction;

const onRender = jest.fn();

SchedulerTracing.unstable_trace("initial_event", performance.now(), () => {
  const interactions = SchedulerTracing.unstable_getCurrent();
  expect(interactions.size).toBe(1);
  interaction = Array.from(interactions)[0];

  ReactDOM.render(
    <React.Profiler id="profiler" onRender={onRender}>
      <Text text="Hi" />
    </React.Profiler>,
    document.createElement("div")
  );
});

expect(interaction).toBeDefined();
expect(onRender).toHaveBeenCalledTimes(1);
expect(onRender.mock.calls[0][6]).toContain(interaction);
```

But if you wrap it in `act`, it fails:
```js
let interaction;

const onRender = jest.fn();

TestUtils.act(() => {
  SchedulerTracing.unstable_trace("initial_event", performance.now(), () => {
    const interactions = SchedulerTracing.unstable_getCurrent();
    expect(interactions.size).toBe(1);
    interaction = Array.from(interactions)[0];

    ReactDOM.render(
      <React.Profiler id="profiler" onRender={onRender}>
        <Text text="Hi" />
      </React.Profiler>,
      document.createElement("div")
    );
  });

  expect(interaction).toBeDefined();
  expect(onRender).toHaveBeenCalledTimes(1);
  expect(onRender.mock.calls[0][6]).toContain(interaction);
});
```

In the latter case, inside of `ReactFiberScheduler`, the interactions `Set` is empty.
https://github.com/facebook/react/blob/2e5d1a8b9e2c29418a27b24306e4c8d5f8681f4f/packages/react-reconciler/src/ReactFiberScheduler.js#L2119-L2126

Note that this does not seem to impact _updates_, only the initial _mount_. I suspect we just need to wrap a callback or something minor.

I came across this while trying to write profiling tests for the new DevTools. (https://github.com/bvaughn/react-devtools-experimental/pull/258)

**Comment by bvaughn** [comment]
Here's a test that removes the `Profiler` from the picture, so the problem is less obscured. This test passes without `act` but fails (the mount case) with it.

```js
let expectedInteraction;

const Component = jest.fn(() => {
  expect(expectedInteraction).toBeDefined();

  const interactions = SchedulerTracing.unstable_getCurrent();
  expect(interactions.size).toBe(1);
  expect(interactions).toContain(expectedInteraction);

  return null;
});

const container = document.createElement('div');
const onRender = jest.fn();

act(() => {
  SchedulerTracing.unstable_trace('mount', performance.now(), () => {
    const interactions = SchedulerTracing.unstable_getCurrent();
    expect(interactions.size).toBe(1);
    expectedInteraction = Array.from(interactions)[0];

    ReactDOM.render(<Component />, container);
  });
});

act(() => {
  SchedulerTracing.unstable_trace('update', performance.now(), () => {
    const interactions = SchedulerTracing.unstable_getCurrent();
    expect(interactions.size).toBe(1);
    expectedInteraction = Array.from(interactions)[0];

    ReactDOM.render(<Component />, container);
  });
});

expect(Component).toHaveBeenCalledTimes(2);
```

**Comment by bvaughn** [comment]
I think this has to do with batching behavior. If we remove `act` and mount the component, when it renders- the expected interactions are in scope. If we use `act` though, the interactions set is empty.

Not sure if this is meaningful yet, but I noticed that in the failing case, the component is being run through the legacy un-batched path when mounting:
https://github.com/facebook/react/blob/2e5d1a8b9e2c29418a27b24306e4c8d5f8681f4f/packages/react-reconciler/src/ReactFiberScheduler.js#L333-L340

**Comment by threepointone** [comment]
Interesting. Can you tell me if the bug persists if you just replace act() with batchedUpdates()?

**Comment by bvaughn** [comment]
Good question. In the middle of stepping through some stuff atm. Will check shortly.

**Comment by bvaughn** [comment]
Yeah, so it's _is_ batched updates that's causing the problem. If a _mount_ happens inside of a batched update, then interaction context is lost. Confirmed by replacing `act` with `unstable_batchedUpdates` in my test.

**Comment by bvaughn** [comment]
Looks like interactions are being lost between `renderRoot` and `performUnitOfWork`

**Comment by bvaughn** [comment]
Think I got it sorted.

**Comment by varsha-teckchandani** [comment]
@bvaughn  how did you fix this issue?

**Comment by bvaughn** [comment]
This was caused by a bug in scheduler that was fixed in #15567. I wouldn't guess it's related to what you're seeing (if you're seeing a bug in a product test).
